README.md
# amWiki 轻文库

![amWiki logo](https://amwiki.xf09.net/docs/assets/logo.png)  
amWiki 是一款由 JS 开发、依赖 Atom 或 Nodejs-Npm 的 Markdown 轻量级前端化开源文库系统。  
amWiki 致力于让大家可以更简单、更便捷的建设个人和团队文库！  

[[fork amWiki on Github](https://github.com/TevinLi/amWiki)]  

GitHub:  
[![](https://img.shields.io/github/stars/TevinLi/amWiki.svg?style=social&label=Star)](https://github.com/TevinLi/amWiki "GitHub Stars") [![](https://img.shields.io/github/forks/TevinLi/amWiki.svg?style=social&label=Fork)](https://github.com/TevinLi/amWiki "GitHub Forks") [![](https://img.shields.io/github/issues-raw/TevinLi/amWiki.svg)](https://github.com/TevinLi/amWiki "GitHub Open Issues") [![](https://img.shields.io/github/issues-closed-raw/TevinLi/amWiki.svg)](https://github.com/TevinLi/amWiki "GitHub Closed Issues")  
[![GitHub Tag](https://img.shields.io/github/tag/TevinLi/amWiki.svg)](https://github.com/TevinLi/amWiki "GitHub Tag") [![apm](https://img.shields.io/github/license/TevinLi/amWiki.svg)](https://atom.io/packages/amWiki "MIT License")  
APM:  
[![APM Version](https://img.shields.io/apm/v/amWiki.svg)](https://atom.io/packages/amWiki "APM Version") [![APM Downloads](https://img.shields.io/apm/dm/amWiki.svg)](https://atom.io/packages/amWiki "APM Downloads")  
NPM:  
[![NPM Version](https://img.shields.io/npm/v/amwiki.svg)](https://www.npmjs.com/package/amwiki "NPM Version") [![APM Downloads](https://img.shields.io/npm/dt/amwiki.svg)](https://www.npmjs.com/package/amwiki "APM Downloads") [![npm](https://img.shields.io/npm/dm/amwiki.svg)](https://www.npmjs.com/package/amwiki)

<br>

## amWiki 的优势
- 文档系统采用 markdown 语法
- 不用数据库，文档使用 `.md` 格式保存本地文件
- 无需服务端开发，只需支持 http 静态访问网页空间
- 一键创建新的文库
- 自动更新文库导航目录
- 支持多级目录
- 支持截图直接粘帖为本地 png 并插入当前 markdown
- 文档web端自适应显示，适合所有平台
- 支持接口文档自动抓取内容生成简单的ajax测试
- 无需服务端的全文库标题内容搜索与计分排序
- ... (更多内容期待您的发现)

<br>

## 安装 amWiki
amWiki 可以同时在 Atom 编辑器和 nodejs npm 的命令行两个平台工作，两个平台的工作相互独立，但所创建的文库却可以相互共用  
（PS：对这两个平台的依赖都是编辑需求而不是服务器需求，amWiki 创建的文库是纯静态 html，可以布置到任意服务器）

### 作为 Atom 插件安装
1. 下载 Github 开源文本编辑器 [Atom](https://atom.io/ "Atom官网")，并安装  
2. 安装 Atom 插件 amWiki，并在完成后重启 Atom
    - 前往 Github 的 [amWiki版本发布页](https://github.com/TevinLi/amWiki/releases) 下载最新版压缩包，解压到 C:\Users\Administrator\.atom\packages，并将文件夹名 `amWiki-1.x.x` 改为 `amWiki`
    - 或者，Atom 菜单，File -> Setting -> Install -> 搜索 `amWiki` -> 找到 amWiki 并  Install
    - 或者，在 cmd 或终端中命令：`apm install amWiki`（_第三字母 W 大写_）
3. 在 Atom 菜单，File -> `Add Project Folder` 添加一个项目文件夹
4. 在此文件夹下创建一个名为 `config.json` 的文件
5. 在 Atom 菜单，amWiki轻文库 -> 通过“config.json”创建新文库

### 作为 nodejs 全局模块安装
1. 下载 [nodejs](https://nodejs.org/) 并安装
2. 执行命令： `npm install -g amwiki`（_第三字母 w 小写_）
3. cd 到某个文件夹，通过命令 `amwiki create` 创建文库
4. 通过命令 `amwiki help` 查看帮助

<br>

## amWiki 教程 & 文档
更多详细功能与使用介绍请前往: [ [amWiki 官网文档中心：amwiki.org/doc/](http://amwiki.org/doc/) ]  
(amWiki 的文档由项目自身创建与维护)  

<br>

## 更新日志
### 如何获取最新版本信息与动态？
- QQ群 **347125653**
- 如果可以，欢迎直接加我微信  
  ![](http://img.amwiki.org/global/wx.qrcode.150.png)

### 如何查看历史版本日志？
[amWiki 版本更新日志](https://github.com/TevinLi/amWiki/blob/master/CHANGELOG.md "amWiki版本更新日志")


CHANGELOG.md
# amWiki 版本日志

## [v1.2.1](https://github.com/TevinLi/amWiki/releases/tag/v1.2.1)
_2017-08-08_

- 改善 Mac 上点号开头的系统文件带来的诸多问题（例如导出报错等）([#32](https://github.com/TevinLi/amWiki/issues/32)、[#67](https://github.com/TevinLi/amWiki/issues/67))
- 允许在 config 上重新定义 library 路径 ([#47](https://github.com/TevinLi/amWiki/issues/47))
- 文章篇内目录显示优化 ([#33](https://github.com/TevinLi/amWiki/issues/33))
- 文章篇内目录序号识别，已有序号不再增加序号 ([#33](https://github.com/TevinLi/amWiki/issues/33))
- 基础 Markdown 语法扩充，实现文字飘红
- 统一官网和文档中心网址
- 修复几处显示细节

## [v1.2.0](https://github.com/TevinLi/amWiki/releases/tag/v1.2.0)
_2017-07-18_

- 新增本地数据挂载模式，可以直接双击 index 在浏览器中访问整个文库
    - Atom 平台和 Node-Npm 平台挂载数据操作绑定
    - 本地页面挂载数据生成与嵌入模块
    - 页面加载数据，页面打开时立即判断和读取挂载数据
    - 改进搜索，网络不可用时不再用 worker 而是直接搜索
    - disable 本地模式状态下的接口测试入口
    - 隐藏本地模式状态下的重建缓存按钮
- 基础 Markdown 语法扩充，支持 todo list 复选框 ([#56](https://github.com/TevinLi/amWiki/issues/56))
- 终端本地服务器命令新增可关闭 amWiki 索引页的可选参数
- 优化本地服务器容错性
- 搜索面板结果列表点击不再跳转页面
- 导航增加文档文件名转义，文档文件名允许使用语法符号`()[]'"`而不再冲突
- 对带`&`符号的文件名在URL地址上用特殊转义替代，不会再被作为参数截断 ([#64](https://github.com/TevinLi/amWiki/issues/64))
- 修复服务器文件系统为GBK编码时深目录无法读取的bug
- 调整大量注释、调整一些变量名，标记私有与公有

## [v1.1.3](https://github.com/TevinLi/amWiki/releases/tag/v1.1.3)
_2017-06-25_

- 增加大图预览模式第一期 ([#44](https://github.com/TevinLi/amWiki/issues/44))
- Markdown 基础语法扩充，支持定义图片大小与对齐方式 ([#45](https://github.com/TevinLi/amWiki/issues/45))
- 允许首页的文件名通过增加 `home-` 前缀来自定义
- 修复图片拖拽丢失扩展名的问题 ([#58](https://github.com/TevinLi/amWiki/issues/58))
- 更改 highlight.js 为完整版，支持 176 种语言语法高亮
- 修复测试面板切换文档时的一个显示问题

## [v1.1.2](https://github.com/TevinLi/amWiki/releases/tag/v1.1.2)
_2017-06-03_

- 允许在 config.json 上使用 imports 将额外的 css 和 js 文件嵌入页面 ([#43](https://github.com/TevinLi/amWiki/issues/43))
- 允许本地 server 多开，端口被占用时累加端口号 ([#46](https://github.com/TevinLi/amWiki/issues/46))
- 新增本地 server 索引页，本地根 url 不再 404
- 修复创建时取消创建操作的一个问题
- 调整一些文案

## [v1.1.1](https://github.com/TevinLi/amWiki/releases/tag/v1.1.1)
_2017-05-21_

- 紧急修复 win 平台 npm 创建报错的问题

## [v1.1.0](https://github.com/TevinLi/amWiki/releases/tag/v1.1.0)
_2017-05-20_

- 实现无限级导航 ([#5](https://github.com/TevinLi/amWiki/issues/5)、[#30](https://github.com/TevinLi/amWiki/pull/30))
    - 实现导航水平滚动条，调和水平垂直两滚动条冲突
    - 实现 PC、Mobi 端导航栏无限级梯度的层级显示
    - 实现目录无限级深度树形与列表数据读取
    - 实现无限级导航 Markdown 生成
    - 实现无限级重复的排序 ID 检查
    - 实现无限级图片粘贴、文件拖拽
- 优化工作端查找 library 文件夹
- 实现拖拽文件文件名汉字转拼音
- 修复导航初始化时展开的一个问题

---

### v1.0.6
_2017-05-05_

- 修复未开启测试模块的文库 web 端无内容显示的问题
- 修复导航未展开目录时显示不全只有一边的问题
- 修复上一版本 npm 失误的发布造成 Linux 平台命令无效的问题
- 移动端点击搜索按钮时关闭导航层 ([#40](https://github.com/TevinLi/amWiki/issues/40))

### v1.0.5
_2017-04-28_

- 导航筛选功能增强
    - 文件夹匹配
    - 筛选多个词
    - 忽略筛选词大小写
    - 筛选条件非主动删除不消失
    - 使用递归进行筛选，为无限级做准备 ([#5](https://github.com/TevinLi/amWiki/issues/5))
- 修复web端在移动端点击导航导航不自动收起的问题 ([#40](https://github.com/TevinLi/amWiki/issues/40))
- 修复web端测试面板在不关闭的条件下跳转文档后的一个显示问题
- 菜单栏增加 `从剪切板粘贴截图` 项，调和与其他插件快捷键冲突 ([#41](https://github.com/TevinLi/amWiki/issues/41))

### v1.0.4
_2017-04-22_

- 紧急修复命令行本地服务器不能启动的问题

### v1.0.3
_2017-04-22_

- 基础 Markdown 语法扩展，支持 footnote 脚注 ([#37](https://github.com/TevinLi/amWiki/issues/37))
- 修复 Windows 平台下 gitbash 执行命令结束报错的问题
- 修复 Linux 平台下 `/usr/bin/env` 解释器调用失败的问题
- 修复 Linux 平台下 create、version 报错的问题
- 释放快捷键绑定 Ctrl+D、Ctrl+H ([#38](https://github.com/TevinLi/amWiki/issues/38))
- 取消 readme 单篇长文介绍，仅保留安装方法，详细介绍引导至文档中心

### v1.0.2
_2017-04-04_

- 命令行启动服务器允许使用自定义端口参数
- 截图粘贴增强，QQ聊天、历史记录面板可复制单张图片直接粘入文档
- 修复命令行板块创建文库操作 $navigation.md 缺失的问题
- 修复本地 http 服务器 500 页面崩溃的问题
- 修复 web 端打开带 hash 地址链接时，hash 无效的问题
- 修复命令行显示版本号报错问题

### v1.0.1
_2017-03-30_

- 修复 npm 渠道安装失败的问题
- readme 添加 npm icon

### v1.0.0
_2017-03-30_

- [ _祝贺 amWiki 项目一周岁_ ]
- 构建基于 npm 的**命令行板块**，使用命令行可以同样创建维护文库
- 重构原工作端，分离为**构建板块**与**编辑器板块**
- 引入 ES6 异步编程，使用 Generator、yield、Thunk，将编辑器板块的 confirm 转为 confirm2，并与命令行板块的 confirm2 调用形式统一化
- 使用 ES6 异步编程，对构建板块部分回调方法进行重构，优先使用 return 与异步编程而不是 callback
- 使用 const/let 变量、解构赋值、箭头函数、Class 等 ES6 语法润色原工作端代码
- 增加命令行、终端界面信息打印着色、帮助菜单、版本显示
- 规范化部分公共方法注释

---

### v0.8.0
_2016-12-30_

- Web 端页面左右分栏滚动 ([#26](https://github.com/TevinLi/amWiki/issues/26))
- 工作端解除部分判断对 F12 造成的限制 ([#27](https://github.com/TevinLi/amWiki/issues/27))

---

### v0.7.7
_2016-12-19_  

- 全文库搜索，增加大小写忽略 ([#24](https://github.com/TevinLi/amWiki/issues/24))
- 修复导出后，在 github wiki 页内目录无法跳转的问题 ([#20](https://github.com/TevinLi/amWiki/issues/20#issuecomment-266927563))
- 修复导出时，html 格式的 img src 相对地址未转换问题 ([#23](https://github.com/TevinLi/amWiki/issues/23))
- 修复导出时，链接与 a 标签相对地址未转换问题 ([#23](https://github.com/TevinLi/amWiki/issues/23))

### v0.7.6
_2016-12-11_  

- 增加文库导出模块，可以将文库导出为 Github Wiki 项目 (**[#20](https://github.com/TevinLi/amWiki/issues/20)**)
- 允许 `.md` 文件文件名带空格 ([#19](https://github.com/TevinLi/amWiki/issues/19))
- Web 移动端点击页内目录以外的区域将折叠目录 ([#22](https://github.com/TevinLi/amWiki/issues/22))
- 修复某些情况下跳页面不返回顶部的问题 ([#21](https://github.com/TevinLi/amWiki/issues/21))

### v0.7.5
_2016-12-03_  
(none)

### v0.7.4
_2016-12-03_

- 增加 Web 端右侧目录悬浮窗 ([#15](https://github.com/TevinLi/amWiki/issues/15))
- 增加对 `[TOC]` 目录生成标签的支持 ([#15](https://github.com/TevinLi/amWiki/issues/15))
- 修复工作端抽取页内目录含单双引号、圆括号时无法正确解析的问题 ([#16](https://github.com/TevinLi/amWiki/pull/16))
- 修复工作端抽取页内目录无法跳转的问题 ([#15](https://github.com/TevinLi/amWiki/issues/15))
- 修复工作端抽取页内目录时内容为空报错的问题 ([#13](https://github.com/TevinLi/amWiki/issues/13), [#14](https://github.com/TevinLi/amWiki/pull/14))
- 修复 Web 端浏览器后退按钮报错问题
- Atom 菜单栏添加 “关于amWiki” 按钮

### v0.7.3
_2016-10-28_

- 修复部分atom下插件不能正常工作的问题

### v0.7.2
_2016-10-17_

- 新增搜索接口地址单独匹配并参与得分排序
- 新增搜索结果接口类文档专属标记
- 允许搜索框直接回车启动搜索 ([#12](https://github.com/TevinLi/amWiki/issues/12))
- Atom菜单微调整

### v0.7.1
_2016-10-10_

- 紧急修复 0.7.0 版本创建的 wiki 无法打开的问题

### v0.7.0
_2016-10-10_

- 新增全库搜索模块 ([#7](https://github.com/TevinLi/amWiki/issues/7))
    - 搜索worker子进程管理
    - 文档预处理实现
    - 关键词匹配得分排序机制设计
    - 搜索面板与基本操作
- 新增更新所有文档缓存
- 新增缓存统计
- 缓存模块其他优化
- 部分命名调整

---

### v0.6.5
_2016-10-04_  

- 新增一批基础说明文档
    - amWiki轻文库简介
    - amWiki功能导图
    - 如何开始一个新amWiki轻文库
    - 如何编辑amWiki轻文库

lib/creator.js
/**
 * @desc amWiki 工作端·输出web静态资源模块
 * @author 耀轩之
 * @copyright 在Tevin的原基础上修改所得
 */

var fs = require("fs");
var makeNav = require('./makeNav');
var vscode = require('vscode');

/**
 * @desc 复制文件
 * @param[in] from 待拷贝的文件路径
 * @param[in] to 拷贝后的所在路径
 * @return 无
 */
var copyFile = function (from, to) {
    var encoding = from.indexOf('png') >= 0 ? 'binary' : 'utf-8';
    var file = fs.readFileSync(from, encoding);
    fs.writeFileSync(to, file, encoding);
};

/**
 * @desc 创建amWiki需要的文件夹
 * @param[in] outputPath 待创建新文库的所在目录路径
 * @return 如果返回为true，则创建文件夹成功，否则失败
 */
var createDir = function (outputPath) {
    if (!fs.existsSync(outputPath + 'amWiki/')) {
        fs.mkdirSync(outputPath + 'amWiki/', 0777);
    }
    if (!fs.existsSync(outputPath + 'amWiki/js/')) {
        fs.mkdirSync(outputPath + 'amWiki/js/', 0777);
    }
    if (!fs.existsSync(outputPath + 'amWiki/css')) {
        fs.mkdirSync(outputPath + 'amWiki/css', 0777);
    }
    if (!fs.existsSync(outputPath + 'amWiki/images')) {
        fs.mkdirSync(outputPath + 'amWiki/images', 0777);
    }
    if (!fs.existsSync(outputPath + 'library/')) {
        fs.mkdirSync(outputPath + 'library/', 0777);
        if (!fs.existsSync(outputPath + 'library/001-学习amWiki')) {
            fs.mkdirSync(outputPath + 'library/001-学习amWiki', 0777);
        }
        if (!fs.existsSync(outputPath + 'library/001-学习amWiki/05-学习markdown')) {
            fs.mkdirSync(outputPath + 'library/001-学习amWiki/05-学习markdown', 0777);
        }
        if (!fs.existsSync(outputPath + 'library/002-文档示范')) {
            fs.mkdirSync(outputPath + 'library/002-文档示范', 0777);
        }
        return false;
    }
    return true;
};

/**
 * @desc 创建着色色系（颜色变亮变暗不是rgb分别同时增加或减少一个值）
 * @param[in] 待转换为十六进制颜色数值
 * @return 返回转化后的颜色数值
 */
var makeColours = function (color) {
    var colours = {};
    //修复16进制下的1位数
    var lessThan = function (str) {
        if (str <= 9) {
            str = '0' + str;
        } else {
            var num = parseInt('0x' + str);
            if (num >= 10 && num <= 15) {
                str = '0' + str;
            }
        }
        return str;
    };
    //颜色合法检查
    var atBinary16 = /^#([0-9a-fA-F]{3}){1,2}$/.test(color);
    var atRGB = /^(r|R)(g|G)(b|B)(a|A)?\(\d{1,3},\d{1,3},\d{1,3}(,.*?)?\)$/.test(color);
    var atList = /(blue|default)/.test(color);
    if (!atBinary16 && !atRGB && !atList) {
        vscode.window.showWarningMessage('请使用标准16进制颜色、标准RGB颜色、2种特定颜色名指定颜色！本次创建使用默认颜色');
        return makeColours('default');
    }
    //颜色名称转色值
    var colorList = {
        default: '#4296eb',
        blue: '#4296eb'
    };
    if (atList) {
        color = colorList[color];
    }
    //解析颜色为rgb
    var rgb = [];
    if (color.indexOf('#') == 0) {
        rgb[0] = parseInt('0x' + color.substr(1, 2));
        rgb[1] = parseInt('0x' + color.substr(3, 2));
        rgb[2] = parseInt('0x' + color.substr(5, 2));
    } else {
        var rgbStr = color.split('(')[1].split(')')[0].split(',');
        rgb[0] = parseInt(rgbStr[0]);
        rgb[1] = parseInt(rgbStr[1]);
        rgb[2] = parseInt(rgbStr[2]);
    }
    //生成主色base
    colours.base = '#' + [
        lessThan(rgb[0].toString(16)),
        lessThan(rgb[1].toString(16)),
        lessThan(rgb[2].toString(16))
    ].join('');
    //颜色变化量，越小的值变化越大
    var increment = [
        Math.round((255 - rgb[0]) / 3),
        Math.round((255 - rgb[1]) / 3),
        Math.round((255 - rgb[2]) / 3)
    ];
    //生成亮色light
    colours.light = '#' + [
        (rgb[0] + increment[0]).toString(16),
        (rgb[1] + increment[1]).toString(16),
        (rgb[2] + increment[2]).toString(16)
    ].join('');
    //生成暗色dark
    var larger = 0,
        largeIndex = -1;
    var largerFd = function (i) {
        if (increment[i] > larger) {
            larger = increment[i];
            largeIndex = i;
        }
    };
    largerFd(0);
    largerFd(1);
    largerFd(2);
    if (rgb[largeIndex] - larger < 0) {
        //修复加深颜色值为负的情况，最多减到0，其他三色等比例减小
        var reduce = (rgb[largeIndex] - larger) / larger;
        increment[0] += Math.round(increment[0] * reduce);
        increment[1] += Math.round(increment[1] * reduce);
        increment[2] += Math.round(increment[2] * reduce);
    }
    colours.dark = '#' + [
        lessThan((rgb[0] - increment[0]).toString(16)),
        lessThan((rgb[1] - increment[1]).toString(16)),
        lessThan((rgb[2] - increment[2]).toString(16))
    ].join('');
    return colours;
};


module.exports = {
    /**
     * @desc 开始创建
     * @param[in] state 此参数包含了新文库所需要的数据
     * @return 无
     */
    create: function (state) {
        var options = {};
        var editor = vscode.window.activeTextEditor;
        if (!editor) {
            return;
        }
        options.editorPath = editor.document.fileName;
        if (options.editorPath.indexOf('config.json') < 0) {
            vscode.window.showWarningMessage('当前不是"config.json"文件！');
            return;
        }
        //创建路径
        //获得当前extension所在目录
        var extensionPath = state.context.extensionPath;
        //找到默认文库的所在路径
        options.filesPath = extensionPath.replace(/\\/g, '/') + '/files/';
        options.outputPath = options.editorPath.split('config.json')[0].replace(/\\/g, '/');
        //读取配置
        var config = fs.readFileSync(options.editorPath, 'utf-8') || '';
        if (config.length == 0) {
            vscode.window.showWarningMessage('没有读取到任何配置，继续创建么？', '继续创建').then(function (msg) {
                if ('继续创建' === msg) {
                    config = "{}";
                } else {
                    return;
                }
            });
        }
        //解析默认配置
        var parseOk = true;
        try {
            config = JSON.parse(config);
        } catch (e) {
            vscode.window.showErrorMessage('配置解析失败，请检查您的 config.json 是否有严格按 Json 格式书写！\n错误消息：' + e.message);
            parseOk = false;
        }
        if (!parseOk) {
            return;
        }
        config.name = config.name || 'amWiki轻文库系统';  //库名称
        config.version = typeof config.ver == 'string' ? config.ver : 'by Tevin';  //库版本号
        config.logo = config.logo || 'amWiki/images/logo.png';  //logo地址
        config.testing = config.testing || false;  //是否开启接口测试
        config.colour = config.colour ? makeColours(config.colour) : makeColours('#4296eb');  //设置自定义颜色
        //开始创建
        this.buildAt(options, config, function (path) {
            //添加文库
            state.libraryList.push(path);
            //更新导航
            makeNav.refresh(options.outputPath + 'library/');
        });
    },
    /**
     * @desc 创建amWiki本地文件
     */
    buildAt: function (options, config, callback) {
        //创建
        fs.readdir(options.outputPath, function (err, files) {
            if (files.length > 1) {
                vscode.window.showWarningMessage('此处已有一些文件或文件夹，是否仍然在此创建amWiki？', '继续创建').then(function (msg) {
                    if ('继续创建' !== msg) {
                        return;
                    }
                });
            }
            //创建index.html
            var indexPage = fs.readFileSync(options.filesPath + 'amWiki.tpl', 'utf-8');
            indexPage = indexPage.replace(/\{\{name\}\}/g, config.name)
                .replace('{{version}}', config.version)
                .replace('{{logo}}', config.logo);
            if (config.testing) {
                var testingTpl = fs.readFileSync(options.filesPath + 'amWiki.testing.tpl', 'utf-8');
                var testingScript = '<script src="amWiki/js/amWiki.testing.js"></script>';
                indexPage = indexPage
                    .replace('{{amWiki.testing.tpl}}', testingTpl)
                    .replace('{{amWiki.testing.js}}', testingScript);
            } else {
                indexPage = indexPage
                    .replace('{{amWiki.testing.tpl}}', '')
                    .replace('{{amWiki.testing.js}}', '');
            }
            fs.writeFileSync(options.outputPath + 'index.html', indexPage, 'utf-8');
            //创建文件夹
            var hasLibrary = createDir(options.outputPath);
            //创建amWiki.css
            var wikiCss = fs.readFileSync(options.filesPath + 'amWiki.css', 'utf-8');
            if (config.testing) {
                wikiCss += fs.readFileSync(options.filesPath + 'amWiki.testing.css', 'utf-8');
            }
            wikiCss += fs.readFileSync(options.filesPath + 'amWiki.print.css', 'utf-8');
            wikiCss += fs.readFileSync(options.filesPath + 'amWiki.search.css', 'utf-8');
            wikiCss = wikiCss
                .replace(/@colour-base/g, config.colour.base)
                .replace(/@colour-light/g, config.colour.light)
                .replace(/@colour-dark/g, config.colour.dark);
            fs.writeFileSync(options.outputPath + 'amWiki/css/amWiki.css', wikiCss, 'utf-8');
            //拷贝页面资源
            var fileList = [
                ['markdownbody.github.css', 'amWiki/css/markdownbody.github.css'],
                ['lhjs.github-gist.css', 'amWiki/css/lhjs.github-gist.css'],
                ['gbk.js', 'amWiki/js/gbk.js'],
                ['flowchart.min.js', 'amWiki/js/flowchart.min.js'],
                ['raphael-min.js', 'amWiki/js/raphael-min.js'],
                ['jquery-compat-3.1.0.min.js', 'amWiki/js/jquery-compat-3.1.0.min.js'],
                ['marked.min.js', 'amWiki/js/marked.min.js'],
                ['highlight.min.js', 'amWiki/js/highlight.min.js'],
                ['amWiki.js', 'amWiki/js/amWiki.js'],
                ['amWiki.scrollbar.js', 'amWiki/js/amWiki.scrollbar.js'],
                ['amWiki.testing.js', 'amWiki/js/amWiki.testing.js'],
                ['amWiki.tools.js', 'amWiki/js/amWiki.tools.js'],
                ['amWiki.storage.js', 'amWiki/js/amWiki.storage.js'],
                ['amWiki.docs.js', 'amWiki/js/amWiki.docs.js'],
                ['amWiki.search.js', 'amWiki/js/amWiki.search.js'],
                ['amWiki.search.worker.js', 'amWiki/js/amWiki.search.worker.js'],
                ['icons.svg', 'amWiki/images/icons.svg'],
                ['logo.png', 'amWiki/images/logo.png'],
                ['menubar_bg.png', 'amWiki/images/menubar_bg.png']
            ];
            for (var i = 0; i < fileList.length; i++) {
                copyFile(options.filesPath + fileList[i][0], options.outputPath + fileList[i][1]);
            }
            //如果没有library则复制一套默认文档
            if (!hasLibrary) {
                //首页文档
                var home = fs.readFileSync(options.filesPath + 'doc.home.md', 'utf-8');
                home = home.replace('{{name}}', config.name);
                fs.writeFileSync(options.outputPath + 'library/首页.md', home, 'utf-8');
                //其他页面文档
                var fileList2 = [
                    ['doc.amwiki-introduce.md', 'library/001-学习amWiki/01-amWiki轻文库简介.md'],
                    ['doc.amwiki-mind-map.md', 'library/001-学习amWiki/02-amWiki功能导图.md'],
                    ['doc.amwiki-new.md', 'library/001-学习amWiki/03-如何开始一个新amWiki轻文库.md'],
                    ['doc.amwiki-edit.md', 'library/001-学习amWiki/04-如何编辑amWiki轻文库.md'],
                    ['doc.amwiki-testing.md', 'library/001-学习amWiki/06-使用测试模块测试接口.md'],
                    ['doc.amwiki-test-cross.md', 'library/001-学习amWiki/07-amWiki转接到任意域名进行接口测试.md'],
                    ['doc.md-start.md', 'library/001-学习amWiki/05-学习markdown/01-Markdown快速开始.md'],
                    ['doc.md-high-lighting.md', 'library/001-学习amWiki/05-学习markdown/02-amWiki与语法高亮.md'],
                    ['doc.md-flow-chart.md', 'library/001-学习amWiki/05-学习markdown/03-amWiki与流程图.md'],
                    ['doc.md-atom.md', 'library/001-学习amWiki/05-学习markdown/05-Atom对Markdown的原生支持.md'],
                    ['doc.demo-api.md', 'library/002-文档示范/001-通用API接口文档示例.md'],
                    ['doc.demo-long-article.md', 'library/002-文档示范/002-超长文档页内目录示例.md']
                ];
                for (var j = 0; j < fileList2.length; j++) {
                    copyFile(options.filesPath + fileList2[j][0], options.outputPath + fileList2[j][1]);
                }
            }
            callback && callback(options.outputPath + 'library/');
        });
    }
};